unity-environment/Assets/ML-Agents/Examples/Shared Assets/Scripts/RayPerception.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System.Collections.Generic;
+using UnityEngine;
+
+public class RayPerception : MonoBehaviour
+{
+
+    List<float> perceptionVector = new List<float>();
+
+    public List<float> Percieve(float rayDistance,
+                         float[] rayAngles, string[] detectableObjects,
+                          float startOffset, float endOffset)
+    {
+        perceptionVector.Clear();
+        foreach (float angle in rayAngles)
+        {
+            float noise = 0f;
+            float noisyAngle = angle + Random.Range(-noise, noise);
+            Vector3 position = transform.TransformDirection(
+                GiveCatersian(rayDistance, noisyAngle));
+            position.y = endOffset;
+            Debug.DrawRay(transform.position + new Vector3(0f, startOffset, 0f),
+                          position, Color.black, 0.01f, true);
+            RaycastHit hit;
+            float[] subList = new float[detectableObjects.Length + 2];
+            if (Physics.SphereCast(transform.position +
+                                   new Vector3(0f, startOffset, 0f), 0.5f,
```

Comment:
Isn't it supposed to be 
```Physics.SphereCast(transform.position +
                     new Vector3(0f, startOffset - endOffset, 0f), 0.5f, 
                     position, out hit, rayDistance))
```
?

Reply:
`startOffset` is for where the ray begins (so first argument), and `endOffset` is for where the ray ends (hence being in the 3rd argument.